File: previous_contest/KickStart2020/RoundB_2020/Wandering Robot.py
import math
from sys import stdin

# for interactive problem:
# python interactive_runner.py python local_testing_tool.py 2 -- python interact.py
# The same algorithme can't pass with python
def readarray(tpr):
    return list(map(tpr, stdin.readline().split()))


def readint():
    return int(stdin.readline())

def C(m, n, f):
    return f[m + n] - f[m] - f[n]   

def solve():
    W, H, L, U, R, D = readarray(int)
    #precomputation
    maxn = W + H + 1
    f = [0]
    p = [0]
    for i in range(1, maxn + 1):
        f.append(f[-1] + math.log(i))

    for i in range(1, maxn + 1):
        p.append(p[-1] + math.log(2))

    ans = 0


    if R < W:
        for i in range(U - 1):
            ans += math.exp(C(i, R - 1) - p[i + R])
    
    if D < H:
        for i in range(L - 1):
            ans += math.exp(C(i, D - 1) - p[i + D])
    print(ans)

if __name__ == "__main__":
    t = readint()
    for t_ in range(1, t + 1):
        print("Case #{}: ".format(t_), end="")
        solve()

File: previous_contest/KickStart2020/RoundB_2020/Wandering Robot.cpp
#include <bits/stdc++.h>

using namespace std;

using ll = long long;
using ld = long double;

int t, W, H, L, U, R, D;

ld ans;
constexpr int maxn = 200001;
ld f[maxn], p[maxn];

ld C(int m, int n){
    return f[m + n] - f[m] - f[n];
}

void solve(){
    cout << setprecision(20);
    cin >> W >> H >> L >> U >> R >> D;
    for (int i = 1; i < maxn; i++){
        f[i] = f[i - 1] + log((ld)i);
    }
    for (int i = 1; i < maxn; i++){
        p[i] = p[i - 1] + log((ld)2);
    }
    ans = 0;
    if(R < W) 
        for(int i = 1; i < U; i += 1)
            ans += exp(C(i - 1, R - 1) - p[i + R - 1]);

    if(D < H) 
        for(int i = 1; i < L; i += 1)
            ans += exp(C(i - 1, D - 1) - p[i + D - 1]); 

    cout << ans << endl;
}


int main() {
    ios::sync_with_stdio(false);
    cin >> t;
    for (int t_ = 1; t_ <= t; ++t_) {
        cout << "Case #" << t_ << ": ";
        solve();
    }
    return 0;
}
